File: internal/api/clientgrpc/groups.go
package clientgrpc

import (
	"context"
	"time"

	clientv1 "github.com/torchwooddev/torchwood/genproto/client/v1"
	sharedv1 "github.com/torchwooddev/torchwood/genproto/shared/v1"
	"github.com/torchwooddev/torchwood/internal/app/client"
	"github.com/torchwooddev/torchwood/internal/domain/databases"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
	"google.golang.org/protobuf/types/known/timestamppb"
)

type GroupsService struct {
	clientv1.UnimplementedGroupsServiceServer
	groups *client.Groups
}

func NewGroupsService(groups *client.Groups) *GroupsService {
	return &GroupsService{groups: groups}
}

func (s *GroupsService) CreateGroup(ctx context.Context, req *clientv1.CreateGroupRequest) (*clientv1.Group, error) {
	doc, err := s.groups.CreateGroup(ctx, req.GetName())
	if err != nil {
		return nil, err
	}
	return mapClientGroupDoc(doc), nil
}

func (s *GroupsService) ListGroups(ctx context.Context, req *sharedv1.ListRequest) (*clientv1.ListGroupsResponse, error) {
	docs, total, next, err := s.groups.ListGroups(ctx, databases.Query{
		Queries:   req.GetQueries(),
		PageSize:  req.GetPageSize(),
		PageToken: req.GetPageToken(),
	})
	if err != nil {
		return nil, err
	}
	out := make([]*clientv1.Group, len(docs))
	for i := range docs {
		out[i] = mapClientGroupDoc(&docs[i])
	}
	return &clientv1.ListGroupsResponse{
		Groups: out,
		Meta:   &sharedv1.ListResponseMeta{PageSize: req.GetPageSize(), TotalCount: int32(total), NextPageToken: next},
	}, nil
}

func (s *GroupsService) GetGroup(ctx context.Context, req *clientv1.GetGroupRequest) (*clientv1.Group, error) {
	doc, err := s.groups.GetGroup(ctx, req.GetId())
	if err != nil {
		return nil, err
	}
	if doc == nil {
		return nil, status.Error(codes.NotFound, "group not found")
	}
	return mapClientGroupDoc(doc), nil
}

func (s *GroupsService) DeleteGroup(ctx context.Context, req *clientv1.GetGroupRequest) (*sharedv1.Empty, error) {
	if err := s.groups.DeleteGroup(ctx, req.GetId()); err != nil {
		return nil, err
	}
	return &sharedv1.Empty{}, nil
}

func (s *GroupsService) CreateMembership(ctx context.Context, req *clientv1.CreateMembershipRequest) (*clientv1.Membership, error) {
	doc, err := s.groups.CreateMembership(ctx, req.GetGroupId(), req.GetEmail(), req.GetName(), req.GetRoles())
	if err != nil {
		return nil, err
	}
	return mapClientMembershipDoc(doc), nil
}

func (s *GroupsService) ListMemberships(ctx context.Context, req *clientv1.ListMembershipsRequest) (*clientv1.ListMembershipsResponse, error) {
	docs, err := s.groups.ListMemberships(ctx, req.GetGroupId())
	if err != nil {
		return nil, err
	}
	out := make([]*clientv1.Membership, len(docs))
	for i := range docs {
		out[i] = mapClientMembershipDoc(&docs[i])
	}
	return &clientv1.ListMembershipsResponse{Memberships: out}, nil
}

func (s *GroupsService) UpdateMembershipStatus(ctx context.Context, req *clientv1.UpdateMembershipStatusRequest) (*clientv1.Membership, error) {
	doc, err := s.groups.UpdateMembershipStatus(ctx, req.GetGroupId(), req.GetMembershipId(), req.GetStatus())
	if err != nil {
		return nil, err
	}
	return mapClientMembershipDoc(doc), nil
}

func (s *GroupsService) DeleteMembership(ctx context.Context, req *clientv1.GetMembershipRequest) (*sharedv1.Empty, error) {
	if err := s.groups.DeleteMembership(ctx, req.GetGroupId(), req.GetMembershipId()); err != nil {
		return nil, err
	}
	return &sharedv1.Empty{}, nil
}

func mapClientGroupDoc(doc *databases.Document) *clientv1.Group {
	if doc == nil {
		return nil
	}
	t := &clientv1.Group{
		Id:        doc.ID,
		CreatedAt: timestamppb.New(doc.CreatedAt),
		UpdatedAt: timestamppb.New(doc.UpdatedAt),
	}
	if v, ok := doc.Data["name"].(string); ok {
		t.Name = v
	}
	if v, ok := doc.Data["total"].(float64); ok {
		t.Total = int32(v)
	}
	if v, ok := doc.Data["total"].(int64); ok {
		t.Total = int32(v)
	}
	return t
}

func mapClientMembershipDoc(doc *databases.Document) *clientv1.Membership {
	if doc == nil {
		return nil
	}
	m := &clientv1.Membership{
		Id:        doc.ID,
		CreatedAt: timestamppb.New(doc.CreatedAt),
		UpdatedAt: timestamppb.New(doc.UpdatedAt),
	}
	if v, ok := doc.Data["group_id"].(string); ok {
		m.GroupId = v
	}
	if v, ok := doc.Data["user_id"].(string); ok {
		m.UserId = v
	}
	if v, ok := doc.Data["email"].(string); ok {
		m.Email = v
	}
	if v, ok := doc.Data["name"].(string); ok {
		m.Name = v
	}
	if v, ok := doc.Data["status"].(string); ok {
		m.Status = v
	}
	if arr, ok := doc.Data["roles"].([]any); ok {
		for _, item := range arr {
			if s, ok := item.(string); ok {
				m.Roles = append(m.Roles, s)
			}
		}
	}
	m.InvitedAt = clientDocTimeField(doc.Data, "invited_at")
	m.JoinedAt = clientDocTimeField(doc.Data, "joined_at")
	return m
}

func clientDocTimeField(data map[string]any, key string) *timestamppb.Timestamp {
	v, ok := data[key].(string)
	if !ok || v == "" {
		return nil
	}
	t, err := time.Parse(time.RFC3339Nano, v)
	if err != nil {
		return nil
	}
	return timestamppb.New(t)
}

File: console/src/api/groups.ts
import { api } from "./client";
import type { ApiRequestConfig } from "./client";

export interface Group {
  id: string;
  name: string;
  total: number;
  permissions?: string[];
  created_at: string;
  updated_at: string;
}

export interface Membership {
  id: string;
  group_id: string;
  user_id: string;
  email: string;
  name: string;
  roles: string[];
  status: string;
  invited_at?: string;
  joined_at?: string;
  created_at: string;
  updated_at: string;
}

export async function listGroups(): Promise<Group[]> {
  const res = await api.get<{ groups: Group[] }>("/server/groups");
  return res.data.groups ?? [];
}

export async function getGroup(id: string): Promise<Group> {
  const res = await api.get<Group>(`/server/groups/${id}`);
  return res.data;
}

export async function createGroup(input: { name: string }): Promise<Group> {
  const res = await api.post<Group>("/server/groups", input);
  return res.data;
}

export async function deleteGroup(id: string, config?: ApiRequestConfig): Promise<void> {
  await api.delete(`/server/groups/${id}`, config);
}

export async function getGroupPrefs(id: string): Promise<Record<string, unknown>> {
  const res = await api.get<{ prefs: Record<string, unknown> }>(`/server/groups/${id}/prefs`);
  return res.data.prefs ?? {};
}

export async function updateGroupPrefs(
  id: string,
  prefs: Record<string, unknown>
): Promise<Record<string, unknown>> {
  const res = await api.put<{ prefs: Record<string, unknown> }>(`/server/groups/${id}/prefs`, {
    prefs,
  });
  return res.data.prefs ?? {};
}

export async function listMemberships(groupId: string): Promise<Membership[]> {
  const res = await api.get<{ memberships: Membership[] }>(
    `/server/groups/${groupId}/memberships`
  );
  return res.data.memberships ?? [];
}

export async function createMembership(
  groupId: string,
  input: {
    email?: string;
    user_id?: string;
    name?: string;
    roles?: string[];
    status?: string;
  }
): Promise<Membership> {
  const res = await api.post<Membership>(`/server/groups/${groupId}/memberships`, {
    group_id: groupId,
    ...input,
  });
  return res.data;
}

export async function updateMembership(
  groupId: string,
  membershipId: string,
  input: { roles: string[] }
): Promise<Membership> {
  const res = await api.patch<Membership>(
    `/server/groups/${groupId}/memberships/${membershipId}`,
    {
      group_id: groupId,
      membership_id: membershipId,
      roles: input.roles,
    }
  );
  return res.data;
}

export async function updateMembershipStatus(
  groupId: string,
  membershipId: string,
  status: string
): Promise<Membership> {
  const res = await api.patch<Membership>(
    `/server/groups/${groupId}/memberships/${membershipId}/status`,
    {
      group_id: groupId,
      membership_id: membershipId,
      status,
    }
  );
  return res.data;
}

export async function deleteMembership(
  groupId: string,
  membershipId: string,
  config?: ApiRequestConfig
): Promise<void> {
  await api.delete(`/server/groups/${groupId}/memberships/${membershipId}`, config);
}
